In [1]:
library(glmnet)

Loading required package: Matrix
Loading required package: foreach
Loaded glmnet 2.0-16



In [2]:
crime_data <- read.table("./data 11.1/uscrime.txt", header = TRUE)

In [3]:
head(crime_data)

M    So Ed   Po1  Po2  LF    M.F   Pop NW   U1    U2  Wealth Ineq Prob    
1 15.1 1   9.1  5.8  5.6 0.510  95.0  33 30.1 0.108 4.1 3940   26.1 0.084602
2 14.3 0  11.3 10.3  9.5 0.583 101.2  13 10.2 0.096 3.6 5570   19.4 0.029599
3 14.2 1   8.9  4.5  4.4 0.533  96.9  18 21.9 0.094 3.3 3180   25.0 0.083401
4 13.6 0  12.1 14.9 14.1 0.577  99.4 157  8.0 0.102 3.9 6730   16.7 0.015801
5 14.1 0  12.1 10.9 10.1 0.591  98.5  18  3.0 0.091 2.0 5780   17.4 0.041399
6 12.1 0  11.0 11.8 11.5 0.547  96.4  25  4.4 0.084 2.9 6890   12.6 0.034201
  Time    Crime
1 26.2011  791 
2 25.2999 1635 
3 24.3006  578 
4 29.9012 1969 
5 21.2998 1234 
6 20.9995  682

In [4]:
model <- lm(Crime~., data = crime_data)
summary(model)


Call:
lm(formula = Crime ~ ., data = crime_data)

Residuals:
    Min      1Q  Median      3Q     Max 
-395.74  -98.09   -6.69  112.99  512.67 

Coefficients:
              Estimate Std. Error t value Pr(>|t|)    
(Intercept) -5.984e+03  1.628e+03  -3.675 0.000893 ***
M            8.783e+01  4.171e+01   2.106 0.043443 *  
So          -3.803e+00  1.488e+02  -0.026 0.979765    
Ed           1.883e+02  6.209e+01   3.033 0.004861 ** 
Po1          1.928e+02  1.061e+02   1.817 0.078892 .  
Po2         -1.094e+02  1.175e+02  -0.931 0.358830    
LF          -6.638e+02  1.470e+03  -0.452 0.654654    
M.F          1.741e+01  2.035e+01   0.855 0.398995    
Pop         -7.330e-01  1.290e+00  -0.568 0.573845    
NW           4.204e+00  6.481e+00   0.649 0.521279    
U1          -5.827e+03  4.210e+03  -1.384 0.176238    
U2           1.678e+02  8.234e+01   2.038 0.050161 .  
Wealth       9.617e-02  1.037e-01   0.928 0.360754    
Ineq         7.067e+01  2.272e+01   3.111 0.003983 ** 
Prob        -4.8

#### R-squared is 0.8031 and adjusted R-squared is 0.7078

In [5]:
# Stepwise Regression
stepwise_model <- step(lm(Crime ~ ., data = crime_data))

Start:  AIC=514.65
Crime ~ M + So + Ed + Po1 + Po2 + LF + M.F + Pop + NW + U1 + 
    U2 + Wealth + Ineq + Prob + Time

         Df Sum of Sq     RSS    AIC
- So      1        29 1354974 512.65
- LF      1      8917 1363862 512.96
- Time    1     10304 1365250 513.00
- Pop     1     14122 1369068 513.14
- NW      1     18395 1373341 513.28
- M.F     1     31967 1386913 513.74
- Wealth  1     37613 1392558 513.94
- Po2     1     37919 1392865 513.95
<none>                1354946 514.65
- U1      1     83722 1438668 515.47
- Po1     1    144306 1499252 517.41
- U2      1    181536 1536482 518.56
- M       1    193770 1548716 518.93
- Prob    1    199538 1554484 519.11
- Ed      1    402117 1757063 524.86
- Ineq    1    423031 1777977 525.42

Step:  AIC=512.65
Crime ~ M + Ed + Po1 + Po2 + LF + M.F + Pop + NW + U1 + U2 + 
    Wealth + Ineq + Prob + Time

         Df Sum of Sq     RSS    AIC
- Time    1     10341 1365315 511.01
- LF      1     10878 1365852 511.03
- Pop     1     14127 13691

In [6]:
# Summary of the model
summary(stepwise_model)


Call:
lm(formula = Crime ~ M + Ed + Po1 + M.F + U1 + U2 + Ineq + Prob, 
    data = crime_data)

Residuals:
    Min      1Q  Median      3Q     Max 
-444.70 -111.07    3.03  122.15  483.30 

Coefficients:
            Estimate Std. Error t value Pr(>|t|)    
(Intercept) -6426.10    1194.61  -5.379 4.04e-06 ***
M              93.32      33.50   2.786  0.00828 ** 
Ed            180.12      52.75   3.414  0.00153 ** 
Po1           102.65      15.52   6.613 8.26e-08 ***
M.F            22.34      13.60   1.642  0.10874    
U1          -6086.63    3339.27  -1.823  0.07622 .  
U2            187.35      72.48   2.585  0.01371 *  
Ineq           61.33      13.96   4.394 8.63e-05 ***
Prob        -3796.03    1490.65  -2.547  0.01505 *  
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

Residual standard error: 195.5 on 38 degrees of freedom
Multiple R-squared:  0.7888,	Adjusted R-squared:  0.7444 
F-statistic: 17.74 on 8 and 38 DF,  p-value: 1.159e-10


#### With the new model after performing stepwise regression we get a slightly lower R-squared value of 0.7888, but a slightly higher adjusted R-squared value of 0.7444. 

In [7]:

uscrime_scale <- cbind(as.data.frame(scale(crime_data[,1])), 
                      as.data.frame(crime_data[,2]), 
                      as.data.frame(scale(crime_data[,c(3,4,5,6,7,8,9,10,11,12,13,14,15)])),
                      as.data.frame(crime_data[,16]))
 
# Fix the column names
colnames(uscrime_scale) <- colnames(crime_data)
# Display scaled data
head(uscrime_scale)

M          So Ed         Po1        Po2        LF         M.F        
1  0.9886930 1  -1.3085099 -0.9085105 -0.8666988 -1.2667456 -1.12060499
2  0.3521372 0   0.6580587  0.6056737  0.5280852  0.5396568  0.98341752
3  0.2725678 1  -1.4872888 -1.3459415 -1.2958632 -0.6976051 -0.47582390
4 -0.2048491 0   1.3731746  2.1535064  2.1732150  0.3911854  0.37257228
5  0.1929983 0   1.3731746  0.8075649  0.7426673  0.7376187  0.06714965
6 -1.3983912 0   0.3898903  1.1104017  1.2433590 -0.3511718 -0.64550313
  Pop         NW           U1          U2         Wealth     Ineq      
1 -0.09500679  1.943738564  0.69510600  0.8313680 -1.3616094  1.6793638
2 -0.62033844  0.008483424  0.02950365  0.2393332  0.3276683  0.0000000
3 -0.48900552  1.146296747 -0.08143007 -0.1158877 -2.1492481  1.4036474
4  3.16204944 -0.205464381  0.36230482  0.5945541  1.5298536 -0.6767585
5 -0.48900552 -0.691709391 -0.24783066 -1.6551781  0.5453053 -0.5013026
6 -0.30513945 -0.555560788 -0.63609870 -0.5895155  1.6956723 -1.7044289
  Prob       Time        Crime
1  1.6497631 -0.05599367  791 
2 -0.7693365 -0.18315796 1635 
3  1.5969416 -0.32416470  578 
4 -1.3761895  0.46611085 1969 
5 -0.2503580 -0.74759413 1234 
6 -0.5669349 -0.78996812  682

In [8]:
# Run lasso 
uscrime_lasso <- cv.glmnet(x=as.matrix(uscrime_scale[,-16]), y=as.matrix(uscrime_scale$Crime), 
                      alpha=1, nfolds = 5, type.measure="mse", family="gaussian")

# Find the lamba with the smallest cvm
x <- uscrime_lasso$cvm
which(x == min(x))

[1] 48

In [9]:
# The lamba for the smallest cvm
uscrime_lasso$lambda[which(x == min(x))]

[1] 3.319887

In [10]:
# or
uscrime_lasso$lambda.min

[1] 3.319887

In [11]:
coefficients(uscrime_lasso, s=uscrime_lasso$lambda.min)


16 x 1 sparse Matrix of class "dgCMatrix"
                    1
(Intercept) 894.31532
M           104.46095
So           31.63626
Ed          175.11702
Po1         296.32198
Po2           .      
LF            .      
M.F          52.34557
Pop         -18.94369
NW           14.32038
U1          -71.07987
U2          116.24494
Wealth       53.72200
Ineq        250.17012
Prob        -89.19999
Time          .      

In [12]:
# Now lets make a model with these coefficients 
uscrime_lasso_lm <- lm(Crime ~M+So+Ed+Po1+M.F+Pop+NW+U1+U2+Wealth+Ineq+Prob, 
                    data = uscrime_scale)
summary(uscrime_lasso_lm)


Call:
lm(formula = Crime ~ M + So + Ed + Po1 + M.F + Pop + NW + U1 + 
    U2 + Wealth + Ineq + Prob, data = uscrime_scale)

Residuals:
    Min      1Q  Median      3Q     Max 
-434.18 -107.01   18.55  115.88  470.32 

Coefficients:
            Estimate Std. Error t value Pr(>|t|)    
(Intercept)   897.29      51.91  17.286  < 2e-16 ***
M             112.71      49.35   2.284  0.02876 *  
So             22.89     125.35   0.183  0.85621    
Ed            195.70      62.94   3.109  0.00378 ** 
Po1           293.18      64.99   4.511 7.32e-05 ***
M.F            48.92      48.12   1.017  0.31656    
Pop           -33.25      45.63  -0.729  0.47113    
NW             19.16      57.71   0.332  0.74195    
U1            -89.76      65.68  -1.367  0.18069    
U2            140.78      66.77   2.108  0.04245 *  
Wealth         83.30      95.53   0.872  0.38932    
Ineq          285.77      85.19   3.355  0.00196 ** 
Prob          -92.75      41.12  -2.255  0.03065 *  
---
Signif. codes:  0 '**

In [13]:
# Let's remove factor which p>0.05 and show the new model. 
uscrime_lasso_lm2 <- lm(Crime ~M+ Ed+ Po1+ U2+ Ineq+Prob, data = uscrime_scale)
summary(uscrime_lasso_lm2)


Call:
lm(formula = Crime ~ M + Ed + Po1 + U2 + Ineq + Prob, data = uscrime_scale)

Residuals:
    Min      1Q  Median      3Q     Max 
-470.68  -78.41  -19.68  133.12  556.23 

Coefficients:
            Estimate Std. Error t value Pr(>|t|)    
(Intercept)   905.09      29.27  30.918  < 2e-16 ***
M             131.98      41.85   3.154  0.00305 ** 
Ed            219.79      50.07   4.390 8.07e-05 ***
Po1           341.84      40.87   8.363 2.56e-10 ***
U2             75.47      34.55   2.185  0.03483 *  
Ineq          269.91      55.60   4.855 1.88e-05 ***
Prob          -86.44      34.74  -2.488  0.01711 *  
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

Residual standard error: 200.7 on 40 degrees of freedom
Multiple R-squared:  0.7659,	Adjusted R-squared:  0.7307 
F-statistic: 21.81 on 6 and 40 DF,  p-value: 3.418e-11


In [14]:
# Elastic net model

r2=c()
for (i in 0:10) {
  mod_uscrime_elastic <- cv.glmnet(x=as.matrix(uscrime_scale[,-16]),y=as.matrix(uscrime_scale$Crime),
                            alpha=i/10, nfolds = 5, type.measure="mse",
                            family="gaussian")
  #dev.ratio is the percentage of deviance explained
  #min index for the dev.ratio of the model
  r2 = cbind(r2, mod_uscrime_elastic$glmnet.fit$dev.ratio[which(mod_uscrime_elastic$glmnet.fit$lambda == mod_uscrime_elastic$lambda.min)])
}

#get the best alpha
alpha <- (which.max(r2)-1)/10


#get model with alphas = 0.05
uscrime_elastic <- cv.glmnet(x=as.matrix(uscrime_scale[,-16]), 
                             y=as.matrix(uscrime_scale$Crime), alpha=0.05, 
                             nfolds = 5, type.measure="mse", family="gaussian")

uscrime_elastic_lm = lm(Crime ~M+So+Ed+Po1+Po2+LF+M.F+NW+U1+U2+Wealth+Ineq+Prob
+Time, data = uscrime_scale)
summary(uscrime_elastic_lm)


Call:
lm(formula = Crime ~ M + So + Ed + Po1 + Po2 + LF + M.F + NW + 
    U1 + U2 + Wealth + Ineq + Prob + Time, data = uscrime_scale)

Residuals:
    Min      1Q  Median      3Q     Max 
-380.91 -101.89  -14.77  110.87  505.40 

Coefficients:
            Estimate Std. Error t value Pr(>|t|)    
(Intercept)  906.483     58.484  15.500  < 2e-16 ***
M            112.837     51.691   2.183  0.03649 *  
So            -4.105    147.172  -0.028  0.97792    
Ed           211.246     68.713   3.074  0.00429 ** 
Po1          563.337    311.541   1.808  0.07998 .  
Po2         -313.824    324.701  -0.966  0.34104    
LF           -31.702     58.147  -0.545  0.58939    
M.F           64.479     54.722   1.178  0.24737    
NW            44.572     65.892   0.676  0.50362    
U1          -112.728     73.902  -1.525  0.13699    
U2           143.186     68.749   2.083  0.04535 *  
Wealth        87.836     98.588   0.891  0.37961    
Ineq         269.086     86.824   3.099  0.00403 ** 
Prob        -

In [15]:
uscrime_elastic_lm2 = lm(Crime ~M+Ed+U2+Ineq+Prob, data = uscrime_scale)
summary(uscrime_elastic_lm2)


Call:
lm(formula = Crime ~ M + Ed + U2 + Ineq + Prob, data = uscrime_scale)

Residuals:
   Min     1Q Median     3Q    Max 
-478.8 -233.6  -46.5  143.2  797.1 

Coefficients:
            Estimate Std. Error t value Pr(>|t|)    
(Intercept)   905.09      47.94  18.881  < 2e-16 ***
M             107.25      68.36   1.569  0.12437    
Ed            240.18      81.89   2.933  0.00547 ** 
U2            135.14      55.35   2.441  0.01903 *  
Ineq          117.68      86.03   1.368  0.17880    
Prob         -156.82      55.20  -2.841  0.00697 ** 
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

Residual standard error: 328.6 on 41 degrees of freedom
Multiple R-squared:  0.3565,	Adjusted R-squared:  0.278 
F-statistic: 4.542 on 5 and 41 DF,  p-value: 0.002186
